src/core/api_client_parallel.py

```
@@ -18,56 +18,66 @@ def __init__(self, base_url: str, api_keys: List[str], requests_per_minute_per_k
-            except asyncio.TimeoutError:
-                self.error_counts[key_index] += 1
-                if retry_attempt == 0 and self.num_keys > 1:
-                    next_key_index = (key_index + 1) % self.num_keys
-                    print(f"      Timeout on key {key_index + 1}, retrying with key {next_key_index + 1}...")
-                    return await self._get_with_key(session, next_key_index, endpoint, params, retry_attempt + 1)
-                raise aiohttp.ClientTimeout(f"Request timeout on key {key_index + 1}")
-            except Exception:
-                self.error_counts[key_index] += 1
-                raise
+        async with self.global_semaphore:  # Global rate limiting
+            async with self.key_locks[key_index]:
```

Ruff (0.12.2)

Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)